Src/Sarif.PatternMatcher/AnalyzeCommand.cs

```diff
@@ -122,36 +122,79 @@ internal static void PushInheritedData(SearchDefinitions definitions, Dictionary
 
         internal static void PushInheritedData(SearchDefinition definition, Dictionary<string, string> sharedStrings)
         {
-            definition.FileNameDenyRegex = PushData(definition.FileNameDenyRegex, sharedStrings);
-            definition.FileNameAllowRegex = PushData(definition.FileNameAllowRegex, sharedStrings);
+            definition.FileNameDenyRegex = PushData(definition.FileNameDenyRegex,
+                                                    definition.SharedStrings,
+                                                    sharedStrings);
+
+            definition.FileNameAllowRegex = PushData(definition.FileNameAllowRegex,
+                                                     definition.SharedStrings,
+                                                     sharedStrings);
 
             foreach (MatchExpression matchExpression in definition.MatchExpressions)
             {
-                matchExpression.FileNameDenyRegex = PushData(matchExpression.FileNameDenyRegex, sharedStrings);
+                matchExpression.FileNameDenyRegex = PushData(matchExpression.FileNameDenyRegex,
+                                                             definition.SharedStrings,
+                                                             sharedStrings);
+
                 matchExpression.FileNameDenyRegex ??= definition.FileNameDenyRegex;
 
-                matchExpression.FileNameAllowRegex = PushData(matchExpression.FileNameAllowRegex, sharedStrings);
+                matchExpression.FileNameAllowRegex = PushData(matchExpression.FileNameAllowRegex,
+                                                             definition.SharedStrings,
+                                                             sharedStrings);
+
                 matchExpression.FileNameAllowRegex ??= definition.FileNameAllowRegex;
 
-                matchExpression.ContentsRegex = PushData(matchExpression.ContentsRegex, sharedStrings);
+                matchExpression.ContentsRegex = PushData(matchExpression.ContentsRegex,
+                                                         definition.SharedStrings,
+                                                         sharedStrings);
+
+                if (matchExpression.Level == 0)
+                {
+                    matchExpression.Level = definition.Level;
+                }
+            }
+        }
+
+        private static string PushData(string text, params Dictionary<string, string>[] sharedStringsDictionaries)
+        {
+            if (text?.Contains("$") != true)
```

Comment:
>Contains [](start = 22, length = 8)

should we move to startswith? Why: that could be a real regex character.

Reply:
Your suggestion won't allow shared strings to be embedded in the regex, e.g.,:

$HttpString$QueryString

Yes, $ can be a real regex character, but its presence (when not used for a shared string) will only provoke an unnecessary processing of the string (looking for variables to expand that don't actually exist).

---
In reply to: [555676556](https://github.com/microsoft/sarif-pattern-matcher/pull/101#discussion_r555676556) [](ancestors = 555676556)